### PR TITLE
Force exit on DYNDNS server failure

### PR DESCRIPTION
**Pull request by adrianrudnik**
Possible fix for #26, only tested with `DYNDNS_SERVER_BIND` set, everything else disabled. Should not affect the other related services.

### CODE REVIEW

**Comment by cromefire** [comment]
@britter Can you check whether this resolves the situation?

**Comment by britter** [comment]
I've applied this change to my homelab, see https://github.com/britter/nix-configuration/commit/6abaa4d7d87a9227512518622e16ce6c3c2cdfa5. This looks better in the sense that the program exists now in case of bind errors. However it exists with 0 exit code. I think a non-zero exit code would be more appropriate in that situation.

File: main.go
Comment (cromefire) [inline main.go:157]: I think the error logging worked better here, where the log can be more specific. Or maybe log in both places? Once for the error once for stopping from that error.

Comment (adrianrudnik) [reply main.go:157]: My reason moving it was to allow for other things to use the context as well, the cause must be an `error` so it needs a log anyway if the context gets "done".

We could wrap the `ListenAndServe()` error and escalate it, if you want the additional information, like

```go
	go func() {
		err := s.ListenAndServe()
		cancel(errors.Join(errors.New("http server error"), err))
	}()
```

but I'm fine with any approach

Comment (cromefire) [reply main.go:157]: Yeah that could make sense, so the error source isn't lost.

**Comment by adrianrudnik** [comment]
Maybe this way. It's not the prettiest output with slog

```
2024/10/04 14:14:58 ERROR Context closed error="http server error\nlisten tcp :8080: bind: address already in use"
```

but contains everything and could be `errors.Unwrap`ed later if more specific handling of the error would be necessary.

**Review by cromefire — APPROVED** [review]

**Comment by cromefire** [comment]
@britter can you check again that it properly exits now? After that I'll merge it and if the other improvements are there as well, I'll also see that there's a release published.

**Comment by britter** [comment]
The latest changes look great to me. When using this app as a systemd unit it's not correctly exiting with non-zero exit code causing it to go into state `failed` which makes it way easier to spot errors.

Thanks a ton for adding this so quickly!

**Pull request merged**
